### PR TITLE
Pretrained embeddings

### PR DESCRIPTION
**Pull request by VIGS25**
This PR adds in:
a) Support for loading pretrained or new embeddings & embedding args
b) Name changes (input -> inputs, ClueClassificator -> ClueClassifier)
c) Corresponding argument changes where needed

### CODE REVIEW

File: supervised/embeddings.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+from __future__ import print_function, unicode_literals, division
+
+import gensim
+import torch
+import torch.nn as nn
+from torchnlp.word_to_vector import GloVe
+import os
+
+LOAD_FUNCS = {'word2vec': load_word2vec,
```
Comment (oskopek) [inline supervised/embeddings.py]: If you change the naming, you don't need this. Just call the `'load_' + embed_type` function.

Comment (oskopek) [reply supervised/embeddings.py]: Also, this is pretty bad coding practice, but meh, it's Python :D Whatever is shorter works, and this is short.

File: supervised/embeddings.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+def load_pretrained(path_to_file, embed_type='word2vec', freeze=False, binary=True):
+	"""Loads pretrained embeddings and returns embeddings and word to index"""
+	if not os.path.exists(path_to_file):
+		raise FileNotFoundError('File does not exist. Please download embeddings and try again.')
+
+	try:
+		load_func = LOAD_FUNCS[embed_type]
+		return load_func(path_to_file, freeze=freeze, binary=binary)
+
+	except KeyError:
```
Comment (oskopek) [inline supervised/embeddings.py]: Remove this and just let it fail. Also, we don't need LOAD_FUNCS at all, see above. It's experimental code, after all.

File: supervised/embeddings.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	:returns: embedding, word_to_index
+	"""
+	path_to_file = embedding_args.get('path_to_file', None)
+	freeze = embedding_args.get('freeze', False)
+
+	if path_to_file is not None:
+		embedding_type = embedding_args.get('embed_type')
+		# TODO (vsomnath): Change the default argument binary
+		return load_pretrained(path_to_file, embed_type=embedding_type, freeze=freeze, binary=True)
+
```
Comment (oskopek) [inline supervised/embeddings.py]: Remove newline

File: supervised/embeddings.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	""" Either loads pretrained embeddings or initializes random embeddings 
+	
+	:param: embedding_args: dictionary containing embedding related arguments
+	:returns: embedding, word_to_index
+	"""
+	path_to_file = embedding_args.get('path_to_file', None)
+	freeze = embedding_args.get('freeze', False)
+
+	if path_to_file is not None:
+		embedding_type = embedding_args.get('embed_type')
```
Comment (oskopek) [inline supervised/embeddings.py]:
```suggestion
		embed_type = embedding_args.get('embed_type')
```

Consistency

File: supervised/embeddings.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+		# TODO (vsomnath): Change the default argument binary
+		return load_pretrained(path_to_file, embed_type=embedding_type, freeze=freeze, binary=True)
+
+	else:
+		num_embeddings = embedding_args['num_embeddings']
+		embedding_dim = embedding_args['embedding_dim']
+		embedding = nn.Embedding(num_embeddings, embedding_dim)
+		if freeze:
+			embedding.weight.require_grad = False
+
```
Comment (oskopek) [inline supervised/embeddings.py]: remove newline

File: supervised/train.py [hunk 6]
```diff
@@ -152,17 +152,27 @@ def main():
...
     board_vocab = data.Vocab.load_from_file(args.board_words)
     clue_vocab = data.Vocab.load_from_file(args.clue_words)
 
-    encoder = network.EncoderRNN(len(board_vocab), hidden_size).to(device)
+    embedding_args = dict() 
+    if args.embedding_path:
+        embedding_args['path_to_file'] = args.embedding_path
+        embedding_args['embed_type'] = util.embed_type(args.embedding_path)
+
+    if args.freeze_embeddings:
```
Comment (oskopek) [inline supervised/train.py]: remove the if

File: supervised/train.py [hunk 7]
```diff
@@ -152,17 +152,27 @@ def main():
...
     args = parser.parse_args()
     assert args.board_words and args.clue_words
 
     hidden_size = 256
     board_vocab = data.Vocab.load_from_file(args.board_words)
     clue_vocab = data.Vocab.load_from_file(args.clue_words)
 
-    encoder = network.EncoderRNN(len(board_vocab), hidden_size).to(device)
+    embedding_args = dict() 
+    if args.embedding_path:
```
Comment (oskopek) [inline supervised/train.py]: remove the if and add a sensible default value to the arg

File: supervised/util.py [hunk 8]
```diff
@@ -26,3 +26,17 @@ def show_plot(points):
     loc = ticker.MultipleLocator(base=0.2)
     ax.yaxis.set_major_locator(loc)
     plt.plot(points)
+
+def embed_type(embeddings_path):
+    """ Returns embedding type based on file extension
```
Comment (oskopek) [inline supervised/util.py]:
```suggestion
    """Returns embedding type based on file extension.
```

File: supervised/util.py [hunk 9]
```diff
@@ -26,3 +26,17 @@ def show_plot(points):
     loc = ticker.MultipleLocator(base=0.2)
     ax.yaxis.set_major_locator(loc)
     plt.plot(points)
+
+def embed_type(embeddings_path):
+    """ Returns embedding type based on file extension 
+    (WARNING: Crude and needs to be improved)
```
Comment (oskopek) [inline supervised/util.py]:
```suggestion
    WARNING: Crude and needs to be improved.
```

File: supervised/util.py [hunk 10]
```diff
@@ -26,3 +26,17 @@ def show_plot(points):
...
+    """
+    extension = embeddings_path.split('.')[-1]
+    if extension == 'bin':
+        return 'word2vec'
+    elif extension == 'dat':
+        return 'glove'
+    elif extension == 'h5py':
+        return 'elmo'
+    else:
+        raise ValueError('Couldn\'t recognize the extension. Allowed values are bin and dat')
```
Comment (oskopek) [inline supervised/util.py]:
```suggestion
        raise ValueError("Couldn't recognize the extension. Allowed values are bin and dat")
```

File: supervised/embeddings.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+def load_pretrained(path_to_file, embed_type='word2vec', freeze=False, binary=True):
+    """Loads pretrained embeddings and returns embeddings and word to index"""
+    if not os.path.exists(path_to_file):
+        raise FileNotFoundError('File does not exist. Please download embeddings and try again.')
+
+    try:
+        load_func = LOAD_FUNCS[embed_type]
+        return load_func(path_to_file, freeze=freeze, binary=binary)
+
+    except KeyError:
```
Comment (oskopek) [inline supervised/embeddings.py]: Remove this and let it fail

File: supervised/embeddings.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+	:returns: embedding, word_to_index
+	"""
+    path_to_file = embedding_args.get('path_to_file', None)
+    freeze = embedding_args.get('freeze', False)
+
+    if path_to_file is not None:
+        embedding_type = embedding_args.get('embed_type')
+        # TODO (vsomnath): Change the default argument binary
+        return load_pretrained(path_to_file, embed_type=embedding_type, freeze=freeze, binary=True)
+
```
Comment (oskopek) [inline supervised/embeddings.py]: Remove newline

File: supervised/embeddings.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+	
+	:param: embedding_args: dictionary containing embedding related arguments
+	:returns: embedding, word_to_index
+	"""
+    path_to_file = embedding_args.get('path_to_file', None)
+    freeze = embedding_args.get('freeze', False)
+
+    if path_to_file is not None:
+        embedding_type = embedding_args.get('embed_type')
+        # TODO (vsomnath): Change the default argument binary
```
Comment (oskopek) [inline supervised/embeddings.py]: What do you mean by this?

File: supervised/util.py [hunk 14]
```diff
@@ -26,3 +26,18 @@ def show_plot(points):
...
```
Comment (oskopek) [inline supervised/util.py]:
```suggestion
        raise ValueError("Couldn't recognize the extension. Allowed values are 'bin' and 'dat'.")
```

File: supervised/embeddings.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+        return load_pretrained(path_to_file, embed_type=embedding_type, freeze=freeze, binary=True)
+
+    else:
+        num_embeddings = embedding_args['num_embeddings']
+        embedding_dim = embedding_args['embedding_dim']
+        embedding = nn.Embedding(num_embeddings, embedding_dim)
```
Comment (oskopek) [inline supervised/embeddings.py]:
```suggestion
        embedding = nn.Embedding(embedding_args['num_embeddings'], embedding_args['embedding_dim'])
```

File: supervised/network.py [hunk 16]
```diff
@@ -6,23 +6,31 @@
...
 
-    def __init__(self, input_size, hidden_size):
+    def __init__(self, input_size, hidden_size, embedding_args={}):
         super(EncoderRNN, self).__init__()
-        self.hidden_size = hidden_size
 
-        self.embedding = nn.Embedding(input_size, hidden_size)
+        # TODO (vsomnath): Pass input_size and hidden_size within embedding args?
+        self.hidden_size = hidden_size
+        self.embedding_args = embedding_args
```
Comment (oskopek) [inline supervised/network.py]: Do we need `self.embedding_args` anywhere? Why not only use `embedding_args`?

File: supervised/network.py [hunk 17]
```diff
@@ -6,23 +6,31 @@
 import torch.nn as nn
 import torch.nn.functional as F
 
+from embeddings import setup_embeddings
```
Comment (oskopek) [inline supervised/network.py]:
```suggestion
from .embeddings import setup_embeddings
```

File: supervised/network.py [hunk 18]
```diff
@@ -6,23 +6,31 @@
...
 device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
 ATTN_MAX_LENGTH = 25  # Maximum number of cards on the board.
 
-# TODO: Find out how to load pretrained embeddings (possibly detach them from training)
-
 
 class EncoderRNN(nn.Module):
 
-    def __init__(self, input_size, hidden_size):
+    def __init__(self, input_size, hidden_size, embedding_args={}):
```
Comment (oskopek) [inline supervised/network.py]: I don't like the embedding_args thing. Lets pass in the embedded tokens directly to these classes? Or lets pass in the initialized `nn.Embedding` module? Or lets pass in the arguments individually? cc @jendelel

File: supervised/embeddings.py [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+def load_GloVe(path_to_file, freeze=False, binary=True):
+	"""Loads pretrained GloVe embeddings"""
+	model_dict = GloVe(name='840B', dim=300)
+	embedding_vectors = list(model_dict.values())
+	embedding_vectors = Torch.FloatTensor(embedding_vectors)
+	word_to_index = {word: i for i, word in enumerate(list(model_dict.keys()))}
+	embedding = nn.Embedding.from_pretrained(embedding_vectors, freeze=freeze)
+	return vectors, word_to_index
+
+def load_ElMo(path_to_file, freeze=False, binary=True):
```
[the remark this answers is not in the document]
Comment (VIGS25) [reply supervised/embeddings.py]: Done

File: supervised/network.py (hunk 18 above)
Comment (VIGS25) [reply supervised/network.py]: I think passing embedded tokens is better.
With the current setup, embedding args just allows you to use random/pretrained directly. But some words could be missing and then with pretrained embeddings, we might need to add in stuff for missing words into a separate function like `_embedding`

File: supervised/train.py (hunk 7 above)
Comment (VIGS25) [reply supervised/train.py]: Like ' ' ?

File: supervised/util.py (hunk 8 above)
Comment (VIGS25) [reply supervised/util.py]: This has been deleted now.
`load_glove` does not use path to file anyway, the Pytorch-NLP takes care of it.

File: supervised/embeddings.py (hunk 11 above)
Comment (VIGS25) [reply supervised/embeddings.py]: Done

File: supervised/embeddings.py [hunk 20]
```diff
@@ -0,0 +1,73 @@
+""" Processing embeddings in PyTorch"""
+
+from __future__ import print_function, unicode_literals, division
+
+import gensim
+import torch
+import torch.nn as nn
+from torchnlp.word_to_vector import GloVe
```
[the remark this answers is not in the document]
Comment (VIGS25) [reply supervised/embeddings.py]: Done

File: supervised/embeddings.py [hunk 21]
```diff
@@ -0,0 +1,73 @@
+""" Processing embeddings in PyTorch"""
+
+from __future__ import print_function, unicode_literals, division
+
+import gensim
```
[the remark this answers is not in the document]
Comment (VIGS25) [reply supervised/embeddings.py]: Done

File: supervised/embeddings.py [hunk 22]
```diff
@@ -0,0 +1,73 @@
+""" Processing embeddings in PyTorch"""
```
[the remark this answers is not in the document]
Comment (VIGS25) [reply supervised/embeddings.py]: Done, sorry the spacing thing was a Sublime issue.

**Pull request closed without merging**